tree_unique_args/src/loop_strength_reduction.egg

```diff
@@ -0,0 +1,111 @@
+(ruleset loop-strength-reduction)
+
+;; Change multiplication of loop variable to addition to an accumulator
+;;
+;; ORIGINAL
+;; a = 0
+;; c = 3
+;; for  i = 0 to n:
+;;     a = i * c
+;;
+;; OPTIMIZED
+;; a = 0
+;; c = 3
+;; d = 0
+;; for i = 0 to n:
+;;     a = d
+;;     d += c
+
+; Finds invariants/constants within a loop.
+; Columns: loop; value of invariant in inputs; value of invariant in outputs
+; TODO: once everything else is working, change this to just use an ID to ref the loop
+(relation Inv (Expr Expr Expr))
+(rule (
+    (= loop (Loop id (All (Parallel) inputs) (All (Sequential) (Pair pred (All (Parallel) outputs)))))
+    (= (ListExpr-ith outputs i) (Get (Arg id) i)))
```

Comment:
Specific to the way the optimization is currently written, we also need to ensure that `(ListExpr-ith inputs i)` is pure, since `(ListExpr-ith inputs i)` will get duplicated later.

I'm also slightly worried that if `(ListExpr-ith inputs i)` is a very expensive (but pure) operation, this optimization actually makes the performance worse and, more importantly, no optimization rule I can think of can recover the performance (no rule can remove `(Get (Args id) i)` from the loop body).